src/llvm_ir_generator.cpp

```diff
@@ -0,0 +1,807 @@
+
+  // A while statement's predicate is evaluated "before" the body statement,
+  // whereas a do-while statement's predicate is evaluated "after" the body
+  // statement. In the generated code for a while statement, there is an
+  // unconditional jump at the end of the body to jump back to the predicate.
+  // For a do-while statement, it only needs one conditional jump.
+  if (!while_stmt.is_do_while) {
+    builder_->CreateBr(pred_bb);
+    builder_->SetInsertPoint(pred_bb);
+    while_stmt.predicate->Accept(*this);
+    auto predicate = val_recorder.ValOfPrevExpr();
+    builder_->CreateCondBr(predicate, body_bb, end_bb);
+  }
+
+  // Connect entry basic block to body basic block.
+  if (while_stmt.is_do_while) {
+    builder_->CreateBr(body_bb);
+  }
+  builder_->SetInsertPoint(body_bb);
+  label_views_of_jumpable_blocks.push_back({.entry = pred_bb, .exit = end_bb});
+  while_stmt.loop_body->Accept(*this);
+  label_views_of_jumpable_blocks.pop_back();
+  llvm_util_.CreateBrIfNoBrBefore(pred_bb);
+
+  if (while_stmt.is_do_while) {
+    builder_->SetInsertPoint(pred_bb);
+    while_stmt.predicate->Accept(*this);
+    auto predicate = val_recorder.ValOfPrevExpr();
+    builder_->CreateCondBr(predicate, body_bb, end_bb);
+  }
+  builder_->SetInsertPoint(end_bb);
+}
+
+void LLVMIRGenerator::Visit(const ForStmtNode& for_stmt) {
+  auto func = llvm_util_.CurrFunc();
+  auto pred_bb = llvm::BasicBlock::Create(*context_, "for_pred", func);
+  auto body_bb = llvm::BasicBlock::Create(*context_, "for_body", func);
+  auto step_bb = llvm::BasicBlock::Create(*context_, "for_step", func);
+  auto end_bb = llvm::BasicBlock::Create(*context_, "for_end", func);
+
+  for_stmt.loop_init->Accept(*this);
+  builder_->CreateBr(pred_bb);
+  builder_->SetInsertPoint(pred_bb);
+  for_stmt.predicate->Accept(*this);
+  if (!dynamic_cast<NullExprNode*>((for_stmt.predicate).get())) {
+    auto predicate = val_recorder.ValOfPrevExpr();
+    builder_->CreateCondBr(predicate, body_bb, end_bb);
+  }
+
+  builder_->SetInsertPoint(body_bb);
+  label_views_of_jumpable_blocks.push_back({.entry = step_bb, .exit = end_bb});
+  for_stmt.loop_body->Accept(*this);
+  label_views_of_jumpable_blocks.pop_back();
+  llvm_util_.CreateBrIfNoBrBefore(step_bb);
+
+  builder_->SetInsertPoint(step_bb);
+  for_stmt.step->Accept(*this);
+  builder_->CreateBr(pred_bb);
+  builder_->SetInsertPoint(end_bb);
+}
+
+void LLVMIRGenerator::Visit(const ReturnStmtNode& ret_stmt) {
+  ret_stmt.expr->Accept(*this);
+  auto expr = val_recorder.ValOfPrevExpr();
+  builder_->CreateRet(expr);
+}
+
+void LLVMIRGenerator::Visit(const GotoStmtNode& goto_stmt) {
+  llvm::BasicBlock* target_bb = llvm_util_.FindBBWithNameOf(goto_stmt.label);
+
+  if (target_bb) {
+    builder_->CreateBr(target_bb);
+  } else {
+    auto label_bb = llvm::BasicBlock::Create(*context_, goto_stmt.label,
+                                             llvm_util_.CurrFunc());
+    builder_->CreateBr(label_bb);
```

Comment:
Are we creating the basic block here if the target doesn't exist yet? If so, we can reassign the value to `target_bb` instead of declaring a new one. This better conveys the fact that we are creating one.